crates/http/src/server/builder.rs

```diff
@@ -0,0 +1,351 @@
+    /// Enabled by default.
+    pub fn http1(mut self, enable: bool) -> Self {
+        self.enable_http1 = enable;
+        self
+    }
+
+    /// Enable HTTP/1 support.
+    ///
+    /// Enabled by default.
+    pub fn enable_http1(self) -> Self {
+        self.http1(true)
+    }
+
+    /// Disable HTTP/1 support.
+    ///
+    /// Enabled by default.
+    pub fn disable_http1(self) -> Self {
+        self.http1(false)
+    }
+
+    /// Enable or disable HTTP/2 support based on the given bool.
+    ///
+    /// Enabled by default.
+    pub fn http2(mut self, enable: bool) -> Self {
+        self.enable_http2 = enable;
+        self
+    }
+
+    /// Enable HTTP/2 support.
+    ///
+    /// Enabled by default.
+    pub fn enable_http2(self) -> Self {
+        self.http2(true)
+    }
+
+    /// Disable HTTP/2 support.
+    ///
+    /// Enabled by default.
+    pub fn disable_http2(self) -> Self {
+        self.http2(false)
+    }
+
+    /// Enable or disable HTTP/3 support based on the given bool.
+    ///
+    /// Disabled by default.
+    pub fn http3(mut self, enable: bool) -> Self {
+        self.enable_http3 = enable;
+        self
+    }
+
+    /// Enable HTTP/3 support.
+    ///
+    /// Disabled by default.
+    pub fn enable_http3(self) -> Self {
+        self.http3(true)
+    }
+
+    /// Disable HTTP/3 support.
+    ///
+    /// Disabled by default.
+    pub fn disable_http3(self) -> Self {
+        self.http3(false)
+    }
+
+    /// Set the rustls configuration for the server.
+    ///
+    /// This enables TLS support for the server.
+    ///
+    /// Required for HTTP/3 support.
+    pub fn with_rustls(mut self, config: rustls::ServerConfig) -> Self {
+        self.rustls_config = Some(config);
+        self
+    }
+
+    /// Build the [`HttpServer`](crate::HttpServer) from this builder.
+    ///
+    /// Make sure to set the bind address and service factory before calling this method.
+    /// If HTTP/3 support is enabled, the rustls configuration must be set as well.
+    pub fn build(mut self) -> Result<HttpServer<F>, ServerBuilderError> {
+        // https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids
+        if let Some(rustlsconfig) = &mut self.rustls_config {
+            rustlsconfig.alpn_protocols.clear();
+
+            if self.enable_http1 {
+                rustlsconfig.alpn_protocols.push(b"http/1.0".to_vec());
+                rustlsconfig.alpn_protocols.push(b"http/1.1".to_vec());
+            }
+
+            if self.enable_http2 {
+                rustlsconfig.alpn_protocols.push(b"h2".to_vec());
+                rustlsconfig.alpn_protocols.push(b"h2c".to_vec());
+            }
+
+            if self.enable_http3 {
+                rustlsconfig.alpn_protocols.push(b"h3".to_vec());
+            }
+        } else if self.enable_http3 {
+            // HTTP/3 doesn't work without TLS
+            return Err(ServerBuilderError::MissingRustlsConfig);
+        }
```

Comment:
This shouldnt be an error, we should require that TLS be provided before you can enable http3. This should be done via the type system (using a stateful builder pattern)